src/build_log.cc

```diff
@@ -418,3 +419,50 @@ bool BuildLog::Recompact(const string& path, const BuildLogUser& user,
 
   return true;
 }
+
+bool BuildLog::Restat(const StringPiece path,
+                      const DiskInterface& disk_interface,
+                      std::string* const err) {
+  METRIC_RECORD(".ninja_log restat");
+
+  Close();
+  std::string temp_path = path.AsString() + ".restat";
+  FILE* f = fopen(temp_path.c_str(), "wb");
+  if (!f) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  if (fprintf(f, kFileSignature, kCurrentVersion) < 0) {
+    *err = strerror(errno);
+    fclose(f);
+    return false;
+  }
+  for (Entries::iterator i = entries_.begin(); i != entries_.end(); ++i) {
+    const TimeStamp mtime = disk_interface.Stat(i->second->output, err);
+    if (mtime == -1) {
+      fclose(f);
+      return false;
+    }
+    i->second->mtime = mtime;
+
+    if (!WriteEntry(f, *i->second)) {
+      *err = strerror(errno);
+      fclose(f);
+      return false;
+    }
+  }
+
+  fclose(f);
+  if (unlink(path.str_) < 0) {
```

Comment:
Like #1572, shouldn't this tool run on an not-run build still succeed? Alternatively, does `ninja` document the file names that will be used for this tool so that tools can know whether to call this tool or not? What happens if run in a build meant for `subninja` inclusion?

Reply:
Good questions. I will add some documentation.

As for running it on a clean build: CMake knows if it is run on a clean build directory, right? It could not call `ninja -t restat` then. But that wouldn't help for re-running cmake before if one hasn't called ninja yet ... hm ...

Do you plan on displaying the output of `ninja -t restat`? I'm thinking about ignoring a non-existing build_log (i.e. return 0), but still display a warning.

Reply:
> CMake knows if it is run on a clean build directory, right?

It knows if it initialized it, but that doesn't correlate with whether this tool will work or not as you mention. Someone could have removed `.ninja_log` manually too.

> Do you plan on displaying the output of ninja -t restat?

I think it is a plain `fork`/`exec`, so its output is tied to CMake's stdout/stderr. I don't know that we want to suppress any output; this isn't the only reason ninja would ever want to output something.

Reply:
If Ninja output a warning but returned 0 if there's no log, would that work for you? Or would you even prefer no warning so that the output is clean?

Reply:
I forsee that someone would complain to CMake about it (especially if CTest ends up catching it and marking it as a warning on CDash). I think the best I could do is point them to this discussion and say that this is what ninja does and CMake can't blanket silence it because errors would also get suppressed (CMake doing the filtering would be…unfortunate especially since every ninja generator would have to do the same thing. Factor in any tool internationalization (unlikely here, but ninja is not a singleton implementation) and it's basically intractable).